### ISSUE TITLE
Build: Show proper version number for beta releases

### ISSUE THREAD

**Issue by hoffie** [jamulussoftware/jamulus#1079]
**Describe the bug**
When tagging a beta release, the built-in version string will be the standard non-release string.

**To Reproduce**
Tag a beta release such as `r3_7_0beta3`.

**Expected behavior**
The version should be 3.7.0beta3 (haven't found any historic beta tags, but this syntax would be consistent with commit-based build numbers (3.6.2git-XXX)

**Screenshots**
The default non-release logic kicks in:
```
$ git checkout r3_7_0beta3
$ qmake Jamulus.pro 
Project MESSAGE: building version "3.6.2dev-6e630f49" (intermediate in git repository)
```

@nefarius2001 Could you tackle that?

cc @ann0see @pljones

**Comment by pljones** [comment]
Thanks for raising somewhere it's visible! :)

**Comment by pljones** [comment]
I'll flag one (off topic, usual style for me) potential issue...

We don't want a beta deployed on the Central Server that clients check for new releases.  Ideally that would stay on 3.6.2 until 3.70 (release) is released.  But it must _not_ trigger the update available message by having something that looks like a 3.7.0 release deployed to it!

**Comment by nefarius2001** [comment]
can do soon, but don't know when exactly :)

**Comment by nefarius2001** [comment]
WIP: https://github.com/nefarius2001/jamulus/tree/autobuild/releasescript

After some bash-scripting, I found python to be most versatile. This way the release can be created from windows and linux systems alike.
Any objections to python as language for that?

**Comment by pljones** [comment]
> Any objections to python as language

Personally, yes... ;)  Any language more obsessed with spacing than I am must have a problem...

This needs to consider what a new developer has installed on their system.  Every new tool that isn't considered "part of the standard platform for that OS desktop user" needs documenting.

We need to tell Windows how to set up a build environment on top of Windows 10 Home, for example.  So, ideally, the fewer add-ons they need to go find and install the better.

(Same applies on other platforms, it's just that "install developer desktop" on MacOS and Linux goes a _very_ long way and it's a much easier experience - although with plenty of alternative paths on the Linux side.)

**Comment by nefarius2001** [comment]
>  Personally, yes... ;) 
Do you suggest alternatives? I tried pure bash at first, but it's much more effort and not good for windows, too. Does Powershell work on linux?

The release-script itself does not necessarily need to support all platforms - just one that the main developers actually use.

**Comment by pljones** [comment]
Everyone who runs a server should be showing the correct version of that server.

That means everyone who follows an old guide that starts by pulling "master" and then compiling and installing the server.

That means quite a lot of people.  Not just developers.  Mostly on Linux but on MacOS and Windows too.  Non-technical people who just want it to work by following step-by-step guides without being aware the universe just changed around them...

The official builds will be built by Github.

**Comment by hoffie** [comment]
@pljones When reading the [script](https://github.com/nefarius2001/jamulus/commit/3c7307cb0d253a1a567fac2c695cbf594756b6b0) I get the impression that this is really "just" the automation of some parts from our release process. It should only be required when one of the main developers creates an official release. It would not be a requirement for simply building a specific git commit, tag or branch.

**Comment by pljones** [comment]
The requirement here is:
- when, as someone doing
  - `git clone`
  - `make clean; qmake; make`
- I get a server with the correct version number.

And that's _regardless_ of how I do it -- this issue is _not_ about release builds through CI.  It's about random people around the world doing builds off the git master.

If they build a beta, the build should have the beta version.

If they build a release, the build should have the release version.

If they build anything else, it should have a dev version, potentially "dirty" if it's not a commit in jamulussoftware/jamulus.

**Comment by hoffie** [comment]
I agree and I think all of those requirements are satisfied. @nefarius2001's script is intended to automate parts of the release process, AFAIU. This will not touch the build logic which runs on each build. Version numbers will still be sourced from Jamulus.pro.

It's about setting the proper version numbers when creating a beta or final release as a main developer.

If people build a beta, they will have to get the beta sources as a tarball, zip or from git. This will already contain the result of what @nefarius2001's script does. No need to run this script in the build process.

Same for releases.

**Comment by nefarius2001** [comment]
here is the flow that I'd propose as beta/release tagging process:
all normal commits have a "dev" in their version in the .pro file
you want to create a beta (neta or full):
0) << if you build here, you get a "dev" with hash appended
1) make a commit that changes the version in the .pro file from "dev" to "beta"/""
2) tag that commit << if you build in this state, you get proper beta or release number
3) make a commit that changes the version in the .pro file back to "dev" << if you build here, you get a "dev" with hash appended
(4) all intermediate commits again build a "dev" with hash appended << if you build here, you get a "dev" with hash appended

this way:
- if you pull and build from master, you get the very latest (dirty) state, and it is marked dirty by the hash in the version string.
- if you pull a beta-tag and build, you get a beta-release-version
- if you pull a release and build, you get a release-version

@pljones I think I do not yet undersstand the pain point. Where is the "update available" mechanism anchored, where does it get the information of versions are available? Shall it pop up if newer commit exists:
- any newer commit on master
- a newer release or beta release
- a newer release? <<< my clear favorite

Also, if you are truely concerned about people suffering from pulling intermediate stuff from master: Then make master a branch that only hopps from a release to the next release (maybe including betas).  Use a different branch name like "development" for all intermediate that the master has now. (I think thats not good. pull master and you get what you can expect from a master. pull a release, and you get what you expect there

**Comment by pljones** [comment]
> Where is the "update available" mechanism anchored, where does it get the information of versions are available?

This is definitely not build process.  As is my way, when I think of something, I try to write it down before forgetting it.  That means it gets written down whatever and wherever I am - this made me think of deployed build version numbers.

As you asked: the client, on start up, queries what was "jamulus.fischvolk.de:22124" for its version.  If the version returned is "newer" (subject to how the code calculates newer), it flags that a new release is available.  This _should_ only be for release builds.

So, what I want is for merges and tag pushes to jamulussoftware/jamulus/master to update `Jamulus.pro` build version numbers correctly.  That means only a "latest" tag push causes a release build version number, a beta tag push sets the version number to the beta release tag and so on.

Remember - NO ONE can update existing deployed software.  It's logically impossible in this universe.  We MUST therefore retain backward compatibility with the upgrade notification system.

**Comment by nefarius2001** [comment]
Let me check if we have the same understanding of the desired/required behavior for the version string of the program:

Everyone gets a server with the correct version number.
-     git clone; make clean; qmake; make   => version contains "dev" and hash
-     git clone; git checkout master;     make clean; qmake; make   => version contains "dev" and hash
-     git clone; git checkout r_3_6_5;     make clean; qmake; make   => version "3.6.5"
-     git clone; git checkout r_3_7_0beta3;     make clean; qmake; make   => version "3.7.0beta3"

@hoffie @pljones  is this what you are looking for?

**Comment by hoffie** [comment]
> @hoffie @pljones is this what you are looking for?

I'd say yes.
I have not checked all details about the script being developed, but I think it would make sense to take a step back. Maybe it would just be enough to add the proper Jamulus.pro update commit for betas to our release checklist and be done with it?

I'm not sure how much effort we should put into automating our release checklist yet. It is the first time we are building and executing it and I think we may still have to rework it after the release?

**Comment by pljones** [comment]
@nefarius2001 Yes, that looks like -- `r_3_7_0beta3` I don't know if we've even agreed the tagging formats yet for betas, though... this looks like a nice one.

As @hoffie says, it's not blocking 3.7.0 - it's great having what's been done so far!

**Comment by ann0see** [comment]
Adding to the question of the tag: why can't we just use r3.7.0? That would match nicely to the version we put in the Jamulus.pro file and might even be a way to automatically get the version number just from the name of the tag.

**Comment by pljones** [comment]
`r3.7.0` is for the release.  The question was over the beta tag.

**Comment by ann0see** [comment]
Yes. But this question is related: Beta tag: r3.7.0beta1 etc.

**Comment by pljones** [comment]
Actually, the release tags use underscore.  It's been like that for a long time and it would be good to keep to the same format.

So `r3_7_0`.

**Comment by nefarius2001** [comment]
>   It's been like that for a long time and it would be good to keep to the same format.

I agree here, stick to established as long as its no pain. (personally, I prefere things like "release/3.7.0")

**Comment by nefarius2001** [comment]
>  I'm not sure how much effort we should put into automating our release checklist yet.

Not sure either, but I think the best documentation is a readable, commented script :)


I have described my proposal here for discussion: 
https://github.com/jamulussoftware/jamulus/discussions/1119
(interrestingl, mentioning this issue in that discussion did not trigger a notification here)

**Comment by hoffie** [comment]
> I have described my proposal here for discussion:

Thanks! I think there are still some things to be sorted out, so I'd suggest to wait for #1120 before continuing work on this.

**Comment by softins** [comment]
Once 3.7.0 has been released with #1120, we will be able to use beta tags correctly, such as 3.7.1beta1

**Comment by hoffie** [comment]
> Once 3.7.0 has been released with #1120, we will be able to use beta tags correctly, such as 3.7.1beta1

With 3.7.0 being released, I think this issue is unblocked and could be worked on. Moving to backlog again for now.
